src/ITensorAutoHOOT/utils.jl

```diff
@@ -0,0 +1,210 @@
+  elseif i < 52
+    return 'A' + i - 26
+  end
+  return Char(i + 140)
+end
+
+"""
+Retrieve the key from the dictionary that maps AutoHOOT nodes
+to ITensor tensors. Returns Nothing if key not exists.
+"""
+function retrieve_key(dict, value)
+  for (k, v) in dict
+    if v === value
+      return k
+    end
+  end
+  return nothing
+end
+
+"""Compute the computational graph defined in AutoHOOT.
+Parameters
+----------
+outnodes: A list of AutoHOOT einsum nodes
+node_dict: A dictionary mapping AutoHOOT node to ITensor tensor
+Returns
+-------
+A list of ITensor tensors.
+"""
+function compute_graph!(out_nodes, node_dict)
+  topo_order = ad.find_topo_sort(out_nodes)
+  for node in topo_order
+    if haskey(node_dict, node) == false && node.name != "1.0"
+      input_list = []
+      for n in node.inputs
+        if haskey(node_dict, n)
+          push!(input_list, node_dict[n])
+        else
+          # the scalar that can neglect
+          @assert(n.name == "1.0")
+        end
+      end
+      node_dict[node] = contract(input_list)
+    end
+  end
+  return [node_dict[node] for node in out_nodes]
+end
+
+compute_graph(out_nodes, node_dict) = compute_graph!(out_nodes, copy(node_dict))
+
+"""Extract an ITensor network from an input network based on AutoHOOT einsum tree.
+The ITensor input network is defined by the tensors in node_dict.
+Note: this function ONLY extracts network based on the input nodes rather than einstr.
+The output network can be hierarchical.
+Parameters
+----------
+outnode: AutoHOOT einsum node
+node_dict: A dictionary mapping AutoHOOT node to ITensor tensor
+Returns
+-------
+A list representing the ITensor output network.
+Example
+-------
+>>> extract_network(einsum("ab,bc->ac", A, einsum("bd,dc->bc", B, C)),
+                    Dict(A => A_tensor, B => B_tensor, C => C_tensor))
+>>> [A_tensor, [B_tensor, C_tensor]]
+"""
+function extract_network(out_node, node_dict)
+  topo_order = ad.find_topo_sort([out_node])
+  node_dict = copy(node_dict)
+  for node in topo_order
+    if haskey(node_dict, node) == false && node.name != "1.0"
+      input_list = []
+      for n in node.inputs
+        if haskey(node_dict, n)
+          push!(input_list, node_dict[n])
+        else
+          # the scalar that can neglect
+          @assert(n.name == "1.0")
+        end
+      end
+      node_dict[node] = input_list
+    end
+  end
+  return node_dict[out_node]
+end
+
+# perform inner product between two nodes
+function inner(n1, n2)
+  @assert(n1.shape == n2.shape)
+  str = join([get_symbol(i) for i in 1:length(n1.shape)], "")
+  return ad.einsum(str * "," * str * "->", n1, n2)
+end
+
+"""Generate ITensor input network based on AutoHOOT einsum expression
+Note: this function generates a network with NEW indices
+Parameters
+----------
+outnode: AutoHOOT einsum node
+node_dict: A dictionary mapping AutoHOOT node to ITensor tensor
+Returns
+-------
+A list of ITensor tensors
+"""
+function generate_network(out_node, node_dict)
```

Comment:
Is this function used anywhere?

Reply:
no it's kind of a deprecated API, I will remove this

Reply:
Done